source/extensions/filters/network/sni_dynamic_forward_proxy/proxy_filter.cc

```diff
@@ -33,14 +35,23 @@ Network::FilterStatus ProxyFilter::onNewConnection() {
     return Network::FilterStatus::Continue;
   }
 
-  // TODO(lizan): implement circuit breaker in SNI dynamic forward proxy like it is in HTTP:
-  // https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc#L65
+  circuit_breaker_ = config_->cache().canCreateDnsRequest(absl::nullopt);
+
+  if (circuit_breaker_ == nullptr) {
+    ENVOY_CONN_LOG(debug, "pending request overflow", read_callbacks_->connection());
+    read_callbacks_->connection().close(Network::ConnectionCloseType::NoFlush);
+    return Network::FilterStatus::StopIteration;
+  }
 
   uint32_t default_port = config_->port();
 
   auto result = config_->cache().loadDnsCacheEntry(sni, default_port, *this);
```

Comment:
more of a meta question: did we consider centralizing this handling in `loadDnsCacheEntry` instead of requiring callers to handle check for overflow?

Reply:
In normal dynamic forward proxy, we can use dns cache circuit breaker. and cluster circuit breakers. When we use cluster circuit breakers and centralize overflow handling, we must leak cluster information to dns cache impl to increment pending stats. I don't agree with this behavior. So we can't centralize them.